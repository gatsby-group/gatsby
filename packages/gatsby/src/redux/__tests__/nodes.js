const Redux = require(`redux`)
const { actions } = require(`../actions`)
const nodeReducer = require(`../reducers/nodes`)
const nodeTouchedReducer = require(`../reducers/nodes-touched`)

jest.mock(`../../db/nodes`)
jest.mock(`../nodes`)

const store = Redux.createStore(
  Redux.combineReducers({ nodeReducer, nodeTouchedReducer }),
  {}
)
const dispatch = jest.spyOn(store, `dispatch`)

describe(`Create and update nodes`, () => {
  beforeEach(() => {
    store.dispatch({
      type: `DELETE_CACHE`,
    })
    dispatch.mockClear()
  })

  it(`allows creating nodes`, async () => {
    await actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: true,
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const action = dispatch.mock.calls[0][0]
    expect(action).toMatchSnapshot()
    expect(nodeReducer(undefined, action)).toMatchSnapshot()
  })

  it(`allows updating nodes`, async () => {
    await actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: true,
        deep: {
          array: [
            0,
            1,
            {
              boom: true,
            },
          ],
        },
      },
      {
        name: `tests`,
      }
    )(dispatch)
    await actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: false,
        deep: {
          array: [1, 2],
        },
        deep2: {
          boom: `foo`,
        },
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const action = dispatch.mock.calls[0][0]
    const updateAction = dispatch.mock.calls[1][0]
    let state = nodeReducer(undefined, action)
    state = nodeReducer(state, updateAction)
    expect(state.get(`hi`).pickle).toEqual(false)
    expect(state.get(`hi`).deep.array[0]).toEqual(1)
    expect(state.get(`hi`).deep2.boom).toEqual(`foo`)
  })

  it(`nodes that are added are also "touched"`, async () => {
    await actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: true,
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const action = dispatch.mock.calls[0][0]
    let state = nodeTouchedReducer(undefined, action)
    expect(state[`hi`]).toBe(true)
  })

  it(`allows adding fields to nodes`, async () => {
    await actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: true,
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const action = dispatch.mock.calls[0][0]
    let state = nodeReducer(undefined, action)

    const addFieldAction = actions.createNodeField(
      {
        node: state.get(`hi`),
        name: `joy`,
        value: `soul's delight`,
      },
      {
        name: `test`,
      }
    )
    state = nodeReducer(state, addFieldAction)
    expect(state).toMatchSnapshot()
  })

  it(`throws error if a field is updated by a plugin not its owner`, async () => {
    await actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `Test`,
        },
        pickle: true,
      },
      {
        name: `tests`,
      }
    )(dispatch)
    const action = dispatch.mock.calls[0][0]
    let state = nodeReducer(undefined, action)

    const addFieldAction = actions.createNodeField(
      {
        node: state.get(`hi`),
        name: `joy`,
        value: `soul's delight`,
      },
      {
        name: `test`,
      }
    )
    state = nodeReducer(state, addFieldAction)

    function callActionCreator() {
      actions.createNodeField(
        {
          node: state.get(`hi`),
          name: `joy`,
          value: `soul's delight`,
        },
        {
          name: `test2`,
        }
      )
    }
    expect(callActionCreator).toThrowErrorMatchingSnapshot()
  })

  it(`throws error if a node is created by a plugin not its owner`, async () => {
    await actions.createNode(
      {
        id: `hi`,
        children: [],
        parent: `test`,
        internal: {
          contentDigest: `hasdfljds`,
          type: `mineOnly`,
        },
        pickle: true,
      },
      {
        name: `pluginA`,
      }
    )(dispatch)

    function callActionCreator() {
      return actions.createNode(
        {
          id: `hi2`,
          children: [],
          parent: `test`,
          internal: {
            contentDigest: `hasdfljds`,
            type: `mineOnly`,
          },
          pickle: true,
        },
        {
          name: `pluginB`,
        }
      )(dispatch)
    }

    expect(await callActionCreator).toThrowErrorMatchingSnapshot()
  })

  it(`throws error if a node sets a value on "fields"`, async () => {
    function callActionCreator() {
      return actions.createNode(
        {
          id: `hi`,
          children: [],
          parent: `test`,
          fields: {
            test: `I can't do this but I like to test boundaries`,
          },
          internal: {
            contentDigest: `hasdfljds`,
            type: `mineOnly`,
          },
          pickle: true,
        },
        {
          name: `pluginA`,
        }
      )(dispatch)
    }

    expect(await callActionCreator).toThrowErrorMatchingSnapshot()
  })
})
