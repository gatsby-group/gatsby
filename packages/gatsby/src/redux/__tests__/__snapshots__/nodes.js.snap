// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create and update nodes allows adding fields to nodes 1`] = `
Object {
  "5efdeb8d-5d6b-5989-8387-faa6426ba350": Object {
    "children": Array [],
    "fields": Object {
      "joy": "soul's delight",
    },
    "id": "5efdeb8d-5d6b-5989-8387-faa6426ba350",
    "internal": Object {
      "contentDigest": "hasdfljds",
      "fieldOwners": Object {
        "joy": "test",
      },
      "owner": "tests",
      "type": "Test",
    },
    "parent": "test",
    "pickle": true,
  },
}
`;

exports[`Create and update nodes allows creating nodes 1`] = `
Object {
  "payload": Object {
    "children": Array [],
    "id": "5efdeb8d-5d6b-5989-8387-faa6426ba350",
    "internal": Object {
      "contentDigest": "hasdfljds",
      "owner": "tests",
      "type": "Test",
    },
    "parent": "test",
    "pickle": true,
  },
  "plugin": Object {
    "name": "tests",
  },
  "traceId": undefined,
  "type": "CREATE_NODE",
}
`;

exports[`Create and update nodes allows creating nodes 2`] = `
Object {
  "5efdeb8d-5d6b-5989-8387-faa6426ba350": Object {
    "children": Array [],
    "id": "5efdeb8d-5d6b-5989-8387-faa6426ba350",
    "internal": Object {
      "contentDigest": "hasdfljds",
      "owner": "tests",
      "type": "Test",
    },
    "parent": "test",
    "pickle": true,
  },
}
`;

exports[`Create and update nodes throws error if a field is updated by a plugin not its owner 1`] = `
"A plugin tried to update a node field that it doesn't own:

Node id: 5efdeb8d-5d6b-5989-8387-faa6426ba350
Plugin: test2
name: joy
value: soul's delight"
`;

exports[`Create and update nodes throws error if a node is created by a plugin not its owner 1`] = `
"The plugin \\"pluginB\\" created a node of a type owned by another plugin.

        The node type \\"mineOnly\\" is owned by \\"pluginA\\".

        If you copy and pasted code from elsewhere, you'll need to pick a new type name
        for your new node(s).

        The node object passed to \\"createNode\\":

        {
    \\"id\\": \\"852da241-f941-5790-b83e-7a0134a7ef8c\\",
    \\"children\\": [],
    \\"parent\\": \\"test\\",
    \\"internal\\": {
        \\"contentDigest\\": \\"hasdfljds\\",
        \\"type\\": \\"mineOnly\\",
        \\"owner\\": \\"pluginB\\"
    },
    \\"pickle\\": true
}

        The plugin creating the node:

        {
    \\"name\\": \\"pluginB\\"
}"
`;

exports[`Create and update nodes throws error if a node sets a value on "fields" 1`] = `
"Plugins creating nodes can not set data on the reserved field \\"fields\\"
      as this is reserved for plugins which wish to extend your nodes.

      If your plugin didn't add \\"fields\\" you're probably seeing this
      error because you're reusing an old node object.

      Node:

      {
    \\"id\\": \\"b5af0cd2-bfc2-533d-b737-d21f086f4de4\\",
    \\"children\\": [],
    \\"parent\\": \\"test\\",
    \\"fields\\": {
        \\"test\\": \\"I can't do this but I like to test boundaries\\"
    },
    \\"internal\\": {
        \\"contentDigest\\": \\"hasdfljds\\",
        \\"type\\": \\"mineOnly\\",
        \\"owner\\": \\"pluginA\\"
    },
    \\"pickle\\": true
}

      Plugin that created the node:

      {
    \\"name\\": \\"pluginA\\"
}"
`;
